bug-fix/Pet_Profile

okhabr opened the pull request:
Closes #41
- Add SCSS instead of CSS;
- Add modules for components styling;
- Correct typescript interfaces;
- Add redux&saga&typesafe-actions for state management;
- Add back button

okhabr committed:
Fix the following bugs:
- Add SCSS instead of CSS;
- Add modules for components styling;
- Correct typescript interfaces;
- Add redux&saga&typesafe-actions for state management;
- Add back button

okhabr committed:
Fix TravisCI catch errors

enheit commented on src/pages/pet/pet.module.scss:
```diff
@@ -0,0 +1,37 @@
+  padding: 16px 0;
+}
+
+// *{
+//     box-sizing: border-box;
+// }
+// ul {
+//     list-style-type: none;
+//     padding: 0;
+// }
```
No "dead code" inside a commit

okhabr replied:
removed

enheit commented on src/store/sagas/pet.saga.ts:
```diff
@@ -0,0 +1,39 @@
+}
+
+async function fetchPet(id:number) {
+  //For testing
+  // const idExample:string = '5e82203cfbbf2d2d6098ac0f';
```
No "dead code" inside a commit

okhabr replied:
removed

enheit commented on src/store/types.ts:
```diff
@@ -0,0 +1,2 @@
+export const PUT_PET :string = 'PUT_PET';
+export const LOAD_PET :string = 'LOAD_PET';
```
The `: string` type is not required during const assignment;

Not perfect
```
const ten: number = 10;
const user: string = 'User';
```

Perfect
```
const ten = 10;
const user = 'User';
```

okhabr replied:
corrected

okhabr committed:
Remove dead code Lines and redundant types

Pashkaua reviewed (approved):
Good job !

Danik1702 reviewed (approved):
Looks great!

enheit commented on src/pages/pet/components/additional-item/additional-item.component.tsx:
```diff
@@ -0,0 +1,23 @@
+
+export const AdditionalItem: React.FC<AdditionalItemProps> = props => {
+  const [open, setOpen] = useState(false);
+  const toggleText = () => {
+    setOpen(!open);
+  };
```
JFYI

Can be simplified
```JavaScript
const toggleText = () => setOpen(!open);
```

okhabr replied:
done

enheit commented on src/pages/pet/components/additional/additional.component.tsx:
```diff
@@ -0,0 +1,20 @@
+import { AdditionalProps } from '../../props.models';
+import { AdditionalItem } from '../additional-item/additional-item.component';
+
+export const AdditionalComponent: React.FC<AdditionalProps> = props => {
```
I would recommend to rename `AdditionalComponent` to the `AdditionalInformationComponent` or `AdditionalInfoComponent` or similar to it. Current name can be treated as additional component for something.

okhabr replied:
done

enheit commented on src/pages/pet/components/additional/additional.component.tsx:
```diff
@@ -0,0 +1,20 @@
+  const blocksArray = Object.entries(blocks);
+  return (
+    <div>
+      {blocksArray.map((entry, index) => (
```
Also, can be improved as

```JavaScript
{blocksArray.map(([title, content], index) => (
  <AdditionalItem title={title} content={content} key={index} />
 ))}
```

okhabr replied:
done

enheit commented on src/pages/pet/components/back-btn/back-btn.component.tsx:
```diff
@@ -0,0 +1,14 @@
+import { Link } from 'react-router-dom';
+import style from './back-btn.module.scss';
+
+export const BackBtn: React.FC<{}> = () => {
```
Correct me if I wrong, but you are able to not specify type by writing `: React.FC` instead of `: React.FC<{}>`

okhabr replied:
you`re right, corrected

enheit commented on src/pages/pet/components/back-btn/back-btn.component.tsx:
```diff
@@ -0,0 +1,14 @@
+export const BackBtn: React.FC<{}> = () => {
+    return (
+      <div className={style.container}>
+      <Link to="/result" className={style.backButton}>
```
Use constants instead of string literal (e.g. `<Link to={ROUTES.result} className={style.backButton}>`)

okhabr replied:
done

enheit commented on src/pages/pet/components/characteristics/characteristics.component.tsx:
```diff
@@ -0,0 +1,28 @@
+import { Rating } from '../rating/rating.component';
+
+export const CharacteristicsComponent: React.FC<CharacteristicsProps> = props => {
+  const { popularity, training, size, brain, security, childFriendly, agility, problematic } = props.data;
```
It is not required to use object destructing. It is okay to use as you used here, but don't overuse it. I nice to have, but not always.

enheit commented on src/pages/pet/components/characteristics/characteristics.component.tsx:
```diff
@@ -0,0 +1,28 @@
+    <ul>
+      {blocksArray.map((entry, index) => (
+        <li className={style.charItem} key={index}>
+          <span>{entry[0]} </span> <Rating rating={entry[1]} />
```
As I said before, you can use array destructing approach here to improve readability.

okhabr replied:
done

enheit commented on src/pages/pet/components/logo/logo.component.tsx:
```diff
@@ -4,7 +4,7 @@ import logo from './logo.svg';
 import './logo.css';
 
-export const Logo = (): React.ReactElement => {
+export const Logo: React.FC<{}> = () => {
```
Redundant type `React.FC<{}>`, can be simplified to `React.FC`

okhabr replied:
done

enheit commented on src/pages/pet/props.models.tsx:
```diff
@@ -0,0 +1,41 @@
+import { AdditionalInfo, Characteristics, Observations } from './pet.models'
+import {PetProfile} from '../../shared/models';
+
+export interface PetProps {
+    match: {
+      params: {
+        id: number;
+      };
+    };
+}
+
+export interface AdditionalProps {
+    data: AdditionalInfo;
+}
+
+export interface AdditionalItemProps {
+    title: string;
+    content: string[];
+  }
+
+export interface CharacteristicsProps {
+    data: Characteristics;
+  }
+
+export  interface ObservationProps {
+    data: Observations;
+  }
+  
+export interface RatingProps {
+    rating: number;
+  }
+
+  export interface RootState {
+    pet: PetState;
+  }
```
Wrong alignment

okhabr replied:
done

enheit commented on src/store/types.ts:
```diff
@@ -0,0 +1,3 @@
+export const FETCH_REQUEST = 'FETCH_REQUEST';
+export const FETCH_SUCCESS = 'FETCH_SUCCESS';
+export const FETCH_ERROR = 'FETCH_ERROR';
```
This file can be removed in favor of string literals. The `typesafe-actions` has good support and able to track any misspelling inside string literals. I mean, you can declare action types inside your action creators.

For example.

Instead of this
```JavaScript
const increment = createCustomAction(actions.INCREMENT);
```
you can do like this
```JavaScript
const increment = createCustomAction('INCREMENT');
```

okhabr replied:
thanks) done

enheit commented on src/store/root.store.ts:
```diff
@@ -1,10 +1,12 @@
 export const store = createStore(rootReducer, composeWithDevTools( applyMiddleware(saga) ));
 
 saga.run(watchPets);
+saga.run(watchLoadPet);
```
It is bad practice to put all watchers inside the root file (P.S. as well you need to put it inside `root.saga.ts` file instead of `root.store.ts`)

Please contact @IvanBonk to improve this part of code. I showed him the better way of structuring sagas.

okhabr replied:
done

okhabr committed:
Fix as per Roma's comments
 - delete redundant types from store
 - create rootsaga
 - fix alignment
 - rename AdditionalComponent (AdditionalInfoComponent)
 - add destructuring to characteristics, additionalInfo components

okhabr committed:
Merge branch 'develop' into bug-fix/Pet_Profile

enheit reviewed (approved):
Oh my! Look at this contribute... Huge work :)
